### PR TITLE
Skip RemoveUnusedRequestParamRector on override abstract method

### LINKED ISSUE #366: `RemoveUnusedRequestParamRector` causes bad PHP code when using abstracts
# Bug Report

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | last dev-main      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.com/demo/381c0067-b8fa-4c7b-8bb1-80f0737a4799

```php
<?php

namespace Symfony\Bundle\FrameworkBundle\Controller {
    class Controller {}
}
    
namespace Symfony\Component\HttpFoundation {
    class Request {}
    class Response {}
}

namespace App\Controller {
    use Symfony\Bundle\FrameworkBundle\Controller\Controller;
    use Symfony\Component\HttpFoundation\Request;
    use Symfony\Component\HttpFoundation\Response;

    abstract class AbstractController extends Controller
    {
        abstract public function fooAction(Request $request): Response;
    }

    class DemoController extends AbstractController
    {
        public function fooAction(Request $request): Response
        {
            return new Response();
        }

        public function barAction(Request $request): Response
        {
            return new Response();
        }
    }
}
```

### Responsible rules

* `RemoveUnusedRequestParamRector`

## Expected Behavior

If the method is an override then it's parameters should not be altered. This would result in inoperable PHP code.

The `$request` parameter should remain in the `fooAction` otherwise the PHP code is inoperable.

### PR DESCRIPTION
**Pull request by JohJohan**
Fixes #366

### CODE REVIEW

**Comment by johanadivare** [comment]
Fixes #366

**Comment by johanadivare** [comment]
I dont know how to fix the rector failing, that has probably nothing to do with my changes.

The PHPstan fails because undefined method but it does work can i add it to the ignores?

**Comment by TomasVotruba** [comment]
@samsonasik It seems the BetterTokenIterator::getTokens() was removed. Do you have any idea how this happened?

**Comment by samsonasik** [comment]
Latest php-doc-parser already have it, so I removed it as it extends the `getTokens()` in rector-src as extends it, see 

- https://github.com/rectorphp/rector-src/pull/3655
- https://github.com/phpstan/phpdoc-parser/commit/900bd69a93e9bd51a11a0bdc6915c4bd813b25c3#diff-62d19c66cc22434f96724388acd962198a200c0c0ced48aeace88157ae658586

php-doc-parser may need to be updated to at least version `^1.20.1`

**Comment by TomasVotruba** [comment]
Ah, I see now 😁

**Comment by samsonasik** [comment]
Let's restart build.

**Comment by samsonasik** [comment]
I will look into why it not load latest phpdoc-parser, @JohJohan  I may cherry-pick your commit in case I can't push to your branch :)

**Comment by JohJohan** [comment]
> I will look into why it not load latest phpdoc-parser, @JohJohan  I may cherry-pick your commit in case I can't push to your branch :) 

Sounds good, good luck :)

**Comment by samsonasik** [comment]
It seems the autoload is overlapped with phpstan.phar's inside build release 1.10.4 which still uses php-doc-parser pinned to `1.19.1`, see https://github.com/phpstan/phpstan-src/blob/532094bc1b70b462c093bfc17eb47b149ff6cb0e/composer.json#L28

rector-src provide preload.php and preload-split-package.php, that's why it works there, I will check if it can be tweaked here :)

**Comment by samsonasik** [comment]
see https://github.com/rectorphp/rector-src/pull/3692

**Comment by samsonasik** [comment]
Let's restart build ...

**Comment by samsonasik** [comment]
@JohJohan I fixed the rector error :tada: , I will let you fix the phpstan notice part :)

File: phpstan.neon [hunk 1]
```diff
@@ -50,6 +50,10 @@ parameters:
         # known type
         - '#Parameter \#1 \$className of method Rector\\Core\\Reflection\\ReflectionResolver\:\:resolveMethodReflection\(\) expects class\-string, string given#'
 
+        -
+            message: '#Call to an undefined method PHPStan\\Reflection\\ExtendedMethodReflection\:\:isAbstract\(\)#'
```
Comment (samsonasik) [inline phpstan.neon:54]: This should be fixed with set actual object with `isAbstract()` method, you can debug actual object with `dump($variable)`

Comment (johanadivare) [reply phpstan.neon:54]: The actual class is `PHPStan\Reflection\Php\PhpMethodReflection`  which has `isAbstract` but when i do something like `assert($extendedMethodReflection instanceof PHPStan\Reflection\Php\PhpMethodReflection);` i get error ` Class Rector\Symfony\Rector\ClassMethod\PHPStan\Reflection\Php\PhpMethodReflection not found. `

Comment (samsonasik) [reply phpstan.neon:54]: You need to use FQCN, or that will refer current namespace:

```diff
-assert($extendedMethodReflection instanceof PHPStan\Reflection\Php\PhpMethodReflection); 
+assert($extendedMethodReflection instanceof \PHPStan\Reflection\Php\PhpMethodReflection); 
```

Comment (johanadivare) [reply phpstan.neon]: Aah thanks :)

**Review by samsonasik — CHANGES_REQUESTED** [review]
Please update PR title to something like:

```
Skip RemoveUnusedRequestParamRector on override abstract method
```

File: src/Rector/ClassMethod/RemoveUnusedRequestParamRector.php [hunk 2]
```diff
@@ -113,4 +124,34 @@ public function refactor(Node $node): ?Node
...
+            return false;
+        }
+        $reflectionParentClass = $classReflection->getParentClass();
+        if (! $reflectionParentClass instanceof ClassReflection) {
+            return false;
+        }
+        try {
+            $extendedMethodReflection = $reflectionParentClass->getNativeMethod($methodName);
+            assert($extendedMethodReflection instanceof PhpMethodReflection);
+            if (! $extendedMethodReflection->isAbstract()) {
```
Comment (samsonasik) [inline src/Rector/ClassMethod/RemoveUnusedRequestParamRector.php:148]: There is already service to verify it, see

https://github.com/rectorphp/rector-src/blob/5c1be93b80c3f7bfd8bd158ba0a55501a7acaefc/packages/FamilyTree/NodeAnalyzer/ClassChildAnalyzer.php#L52

Comment (johanadivare) [reply src/Rector/ClassMethod/RemoveUnusedRequestParamRector.php]: Nice find, i thought there must have been something but i couldn't find it i keep my private function to do the reflection and checking see `hasAbstractParentClassMethod` is that okay?

Comment (samsonasik) [reply src/Rector/ClassMethod/RemoveUnusedRequestParamRector.php]: That's fine as you need to get `ClassReflection` instance 👍

File: tests/Rector/ClassMethod/RemoveUnusedRequestParamRector/Fixture/controller.php.inc [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+abstract class AbstractController extends Controller
+{
+    abstract public function fooAction(Request $request): Response;
+}
+
+class DemoController extends AbstractController
```
Comment (samsonasik) [inline tests/Rector/ClassMethod/RemoveUnusedRequestParamRector/Fixture/controller.php.inc:14]: It needs to be more specific name so it ease to look on the future, eg:

```suggestion
class DontOverrideParentClassMethodController extends AbstractController
```

File: tests/Rector/ClassMethod/RemoveUnusedRequestParamRector/Fixture/controller.php.inc [hunk 4]
```diff
@@ -0,0 +1,55 @@
+<?php
```
Comment (samsonasik) [inline tests/Rector/ClassMethod/RemoveUnusedRequestParamRector/Fixture/controller.php.inc:1]: please rename fixture file to something like: `do_not_override_parent_abstract_class_method.php.inc`

Comment (johanadivare) [reply tests/Rector/ClassMethod/RemoveUnusedRequestParamRector/Fixture/controller.php.inc:1]: ++

File: tests/Rector/ClassMethod/RemoveUnusedRequestParamRector/Fixture/controller.php.inc (hunk 3 above)
Comment (johanadivare) [reply tests/Rector/ClassMethod/RemoveUnusedRequestParamRector/Fixture/controller.php.inc:14]: ++

**Review by samsonasik — APPROVED** [review]

**Comment by samsonasik** [comment]
Thank you @JohJohan

**Pull request merged**